apps/consumers.py

```diff
@@ -80,6 +80,25 @@ def render_async_tracking_data(self, data_id):
             geojson=render_point_feature(data),
         )
 
+    async def telemetry_data_save(self, event):
+        telemetry_data_json = await database_sync_to_async(self.render_async_telemetry_data)(event['telemetry_data_id'])
+        await self.send(telemetry_data_json)
+
+    def render_async_telemetry_data(self, data_id):
+        data = TelemetryData.objects.filter(id=data_id).first()
+        telemetry_data = {
+            'sensor_id': data.sensor_id,
+            'hardware_id': data.hardware_id,
+            'timestamp': str(data.timestamp),
+            'value': data.value,
+        }
+        return Template(
+            '{"event": "$event", "data": {"shipment_id": "$shipment_id", "feature": $telemetry}}').substitute(
```

Comment:
Is the format rendered here under `data` consistent with the format that comes back from the API?

Reply:
I wasn't sure the best way to return it, as the tracking data also contained the "shipment_id" in the template but not in the actual return. The "feature" is the same as an object returned when retrieving telemetry

Reply:
I think it's probably a good idea to make a serializer that includes shipment_id. And then use the rendering method used for the asyncjob and shipment updates, instead of the approach used for the tracking data.

Reply:
You asked for a review - but I don't think you've resolved this yet

Reply:
Wait, so should the telemetry return view also return it in this way? As that is different from the way the telemetry return is returning it.

Reply:
Yeah, I think you could use a Serializer and JSONRenderer in the Telemetry view to clean that up a little. And then you could reuse that here.

Reply:
So, I'm already using a JSONRenderer in the function, and I originally decided against the serializer as I didn't want the typical json+api format, and instead just a list of telemetry data details. Is there a way to do that with the serializer? Also, I'm not sure if I can use the render in the same way as the shipment_update render as that points to a a view which can't really be done with the telemetry endpoint as that isn't a viewset

Reply:
Yes, you use a drf serializer instead of a drf-json-api serializer. And you won't be able to use the exact same syntax since it's not a viewset but it's the same idea.